Sources/HaCWebsiteLib/Controllers/EventStreamController.swift

```diff
@@ -18,14 +18,14 @@ struct LandingUpdateFeedController {
       category: .video,
       description: "Learn all the things.",
       backgroundColor: "#852503",
-      imageURL: "/static/images/functions_frame.png"
+      imageURL: Assets.publicPath("/images/functions_frame.png")
     ),
     PostCard(
       title: "TCP Throughput",
       category: .video,
       description: "Learn all the things.",
       backgroundColor: "green",
-      imageURL: "/static/images/workshop.jpg"
+      imageURL: Assets.publicPath("/images/workshop.jpg")
```

Comment:
Wouldn't it make sense to just give the `imageUrl `as a string to `PostCard`, and have it wrap it with `Assets.publicPath`, makes code a bit more dry no?
In face, couldn't we implement this as an extension to `String`, and have imageUrl treat it as that type?

Reply:
What happens if we want to link to an external image that we don't host?

`Assetable` is worth thinking about though, but that adds a lot of burden to the implementers of these kinds of components (and we tend to make a lot of them)

Reply:
@Pinpickle If it's an external image, won't the url be of the format `http://blah.blah`, it shouldn't be that difficult to treat those differently right?

Reply:
Yeah that is something you could reasonably check for. However, pattern matching on strings kind of goes against all of this nice type checking we're trying to do. I can't imagine a case at the moment, but I don't see it as unthinkable that we want a URL with a leading slash that isn't an asset. Maybe, for example, we have an endpoint that generates them dynamically.

This issue of burden of implementation is a big one. For this idiom to be useful, every component that has an input of an asset of _any_ kind will need to have this layer. This isn't even possible as our lowest level elements from `HaCTML` can't have it. I can imagine this leading to confusion. This is more to think about as implementors, and more to check for as reviewers

It also tightly couples our UI components to our build and asset system. If we ever wanted to make a library of reusable components, we'd be in trouble.

Reply:
```swift
extension String: Assetable {
  var assetPath: String {
  if isRelativeString(self) {
    return Asset.publicPath(self)
  } else {
    return self
  }
}
```

Reply:
How does this add burden?

Reply:
Just saw your new comment, reading now!

Reply:
I think remembering to do this for anything that can be an asset is just another place where we can make mistakes.

But my bigger worry is that it's impossible to do this in all parts of the codebase. HaCTML elements were one example. Another is where we're not necessarily rendering to HTML (maybe we're rendering JSON or making an API call). Users will need to remember to wrap their assets in `publicPath` sometimes but not in others.

Furthermore, I still believe there could be cases where we'd want to have a leading slash and not have the asset resolver step in.

Reply:
@Pinpickle that's why there's a type system right? imageURL is specifically of type Assetable, because it is an asset. That's the documentation saying that you don't need to wrap it. It can also check if strings are valid URLs and complain if they aren't.

Not having it in HaCTML is fine, because that's basically everything in the rawest form.

And for the cases where you don't want the asset resolver to step in with a leading slash, you wouldn't have that variable be `Assetable` anyway right?

Reply:
The type system is kind of deceptive here, as `String`s are still `Assetable`, so we're not really getting any more type safety.

The down-level component can't know whether the consumer wants to resolve on a leading slash or not so it can't make that decision. We could add another type conforming to `Assetable` that has this behaviour but that is starting to seem like a lot of work.

What's your opinion on the tight coupling issue?

Reply:
I just saw that.  It is a good point. But I don't think makes it too hard to change to something reusable in the future. I believe what makes the codebase easier for us to use in the present is more important, but these tradeoffs are worth considering.

What if we created a new type `Asset` that can be initialised with strings? And why isn't the field `imageUrl` and `String?` type instead of a `URL?` type. The latter is more idiomatic no?

Reply:
I agree that components shouldn't care about how the URL was created or whether it's local or remote, it should just accept a URL and not have to process it at all because this logic isn't specific to the component. 

On `String` vs `URL`, the reasoning here is a tradeoff between safety and productivity.
`URL(string: String)` would give us an _optional_ URL. It is nil if it's an invalid URL (i.e. doesn't conform to the grammar of a URL). In practice this means almost anything works (e.g. `hacttp://donu.ts`) works here. We'd be more interested in protecting against mistyped URLs that lead to nowhere (which this doesn't solve). 

Dealing with the optional from URL would practically mean force unwrapping (crashing at runtime if we mistype a URL). This might not be that useful.

Reply:
@jaredkhan So leave it as it is? :)

Reply:
IMO, yeah. 

We could revisit the URL type safety issue later.
For your interest: https://talk.objc.io/episodes/S01E71-type-safe-file-paths-with-phantom-types